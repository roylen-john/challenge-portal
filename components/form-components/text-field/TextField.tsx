import React from 'react'
import { classNames } from '../../../utils/utils'

export interface iTextFieldProps {
  name: string
  type: 'text' | 'email' | 'number' | 'password'
  label?: string
  placeholder?: string
  error?: boolean
  helperText?: string
  fullWidth?: boolean
  value: string
  onChange: () => void
  onBlur: () => void
}

const TextField = React.forwardRef<HTMLInputElement, iTextFieldProps>(
  (
    {
      name,
      type,
      label,
      placeholder,
      error = false,
      helperText,
      fullWidth,
      value,
      onChange,
      onBlur,
    },
    ref
  ) => {
    return (
      <div className={classNames(fullWidth && 'w-full')}>
        {label && (
          <label
            className="block text-gray-700 text-sm font-bold mb-2"
            htmlFor={name}
          >
            {label}
          </label>
        )}
        <input
          className={classNames(
            'shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline',
            error ? 'border-red-500' : ''
          )}
          id={name}
          name={name}
          type={type}
          placeholder={placeholder}
          onChange={onChange}
          onBlur={onBlur}
          value={value}
          ref={ref}
          role="textbox"
        />
        {helperText && (
          <p
            className={classNames(
              'text-xs italic mt-3',
              error ? 'text-red-500' : ''
            )}
          >
            {helperText}
          </p>
        )}
      </div>
    )
  }
)

export default TextField
