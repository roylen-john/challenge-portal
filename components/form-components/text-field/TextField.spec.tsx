import React from 'react'
import { cleanup, fireEvent, render } from '@testing-library/react'

import TextField, { iTextFieldProps } from './TextField'

const DEFAULT_PROPS: iTextFieldProps = {
  name: 'test_input',
  type: 'text',
  label: 'Test Input',
  placeholder: 'Test Input',
  error: false,
  helperText: 'Test input helper text',
  value: '',
  onChange: () => {
    /** Do nothing */
  },
  onBlur: () => {
    /** Do nothing */
  },
}

describe('Textfield', () => {
  afterEach(cleanup)

  test('renders correctly', () => {
    const { asFragment } = render(<TextField {...DEFAULT_PROPS} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('sets input attributes (name, type) correctly', () => {
    const { getByRole } = render(<TextField {...DEFAULT_PROPS} />)
    const input = getByRole('textbox')
    expect(input).toHaveAttribute('name', DEFAULT_PROPS.name)
    expect(input).toHaveAttribute('type', DEFAULT_PROPS.type)
  })

  test('renders label and helper text correctly', () => {
    const { getByText } = render(<TextField {...DEFAULT_PROPS} />)
    const label = getByText(DEFAULT_PROPS.label)
    const helperText = getByText(DEFAULT_PROPS.helperText)
    expect(label).toBeVisible()
    expect(helperText).toBeVisible()
  })

  test('sets value correctly', () => {
    const { getByRole } = render(<TextField {...DEFAULT_PROPS} />)
    const input = getByRole('textbox')
    expect(input).toHaveValue(DEFAULT_PROPS.value)
  })

  test('should call onChange prop', () => {
    const onChangeMock = jest.fn()
    const { getByRole } = render(
      <TextField {...DEFAULT_PROPS} onChange={onChangeMock} />
    )
    const input = getByRole('textbox')
    fireEvent.change(input, { target: { value: 'test-value' } })
    expect(onChangeMock).toHaveBeenCalled()
  })

  test('should call onBlur prop', () => {
    const onBlurMock = jest.fn()
    const { getByRole } = render(
      <TextField {...DEFAULT_PROPS} onBlur={onBlurMock} />
    )
    const input = getByRole('textbox')
    fireEvent.blur(input)
    expect(onBlurMock).toHaveBeenCalled()
  })
})
