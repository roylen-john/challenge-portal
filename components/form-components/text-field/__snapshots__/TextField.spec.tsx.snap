// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textfield renders correctly 1`] = `
<DocumentFragment>
  <div
    class="w-full"
  >
    <label
      class="block text-gray-700 text-sm font-bold mb-2"
      for="test_input"
    >
      Test Input
    </label>
    <input
      class="shadow appearance-none border rounded w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline"
      id="test_input"
      name="test_input"
      placeholder="Test Input"
      role="textbox"
      type="text"
      value=""
    />
    <p
      class="text-xs italic mt-3"
    >
      Test input helper text
    </p>
  </div>
</DocumentFragment>
`;
